File: Fortran/OptimalConsumption.f90
SUBROUTINE OptimalConsumption(lVb,lc,lh,ls,lHc)

USE Parameters
USE Globals
USE Procedures

IMPLICIT NONE

REAL(8),INTENT(IN)	:: lVb
REAL(8),INTENT(OUT)	:: lc,lh,ls,lHc
REAL(8)	:: llabdisutil,lhmax,lhmin
INTEGER	:: iflag
REAL(8), EXTERNAL :: FnHoursBC

!set lVb = -999.9 to use budget constraint rather than FOC for consumption


IF (lVb > -999.0) THEN !not at stationary point or limit
	IF(NoLaborSupply==1) THEN
		lc = utilfn1inv(lVb)
		lh = 1.0
		ls = gbdrift + lh*gnetwage -lc
		IF(lc>0.0) THEN
			lHc = utilfn(lc) + lVb*ls
		ELSE
			lHc = -1.0e12
		END IF

	ELSE IF	(LaborSupplySep==1) THEN
		IF(ScaleDisutilityIdio==0) lh = (gnetwage*lVb/chi)**(frisch)
		IF(ScaleDisutilityIdio==1) lh = ((gnetwage/gidioprod)*lVb/chi)**(frisch)
		IF(ImposeMaxHours==1)lh = min(lh, 1.0_8)
		IF(ScaleDisutilityIdio==0) llabdisutil = lh**(1.0+1.0/frisch)/(1.0+1.0/frisch)
		IF(ScaleDisutilityIdio==1) llabdisutil = gidioprod * lh**(1.0+1.0/frisch)/(1.0+1.0/frisch)
		lc = utilfn1inv(lVb)
		ls = gbdrift + lh*gnetwage -lc
		IF(lc>0.0) THEN
			lHc = utilfn(lc) - chi*llabdisutil + lVb*ls
		ELSE
			lHc = -1.0e12
		END IF
		
	ELSE IF	(LaborSupplyGHH==1) THEN
		IF(ScaleDisutilityIdio==0) lh = (gnetwage/chi)**(frisch)
		IF(ScaleDisutilityIdio==1) lh = ((gnetwage/gidioprod)/chi)**(frisch)
		IF(ImposeMaxHours==1) lh = min(lh, 1.0_8)
		IF(ScaleDisutilityIdio==0) llabdisutil = lh**(1.0+1.0/frisch)/(1.0+1.0/frisch)
		IF(ScaleDisutilityIdio==1) llabdisutil = gidioprod * lh**(1.0+1.0/frisch)/(1.0+1.0/frisch)
		lc = utilfn1inv(lVb) + chi*llabdisutil
		ls = gbdrift + lh*gnetwage -lc
		IF(lc-chi*llabdisutil>0.0) THEN
			lHc = utilfn(lc-chi*llabdisutil) + lVb*ls
		ELSE
			lHc = -1.0e12
		END IF
		
	END IF	
	
ELSEIF (lVb <= -999.0) THEN !at stationary point or limit
	IF(NoLaborSupply==1) THEN
		lh = 1.0
		lc = gbdrift + lh*gnetwage
		ls = 0.0
		IF(lc>0.0) THEN
			lHc = utilfn(lc)
		ELSE
			lHc = -1.0e12
		END IF

	ELSE IF	(LaborSupplySep==1) THEN		
		ls = 0.0
		lhmin = max(0.0_8,-gbdrift/gnetwage+1.0e-5)
		IF(ImposeMaxHours==1) lhmax = 1.0
		IF(ImposeMaxHours==0) lhmax = 100.0
		IF (FnHoursBC(lhmin) >= 0.0) lh = lhmin
		IF (FnHoursBC(lhmax) <= 0.0) lh = lhmax
		IF (FnHoursBC(lhmin)<0.0 .and. FnHoursBC(lhmax)>0.0) THEN		
			CALL rtsec(FnHoursBC,lhmin,lhmax,1.0e-8_8,lh,iflag)
			IF(iflag<0) CALL rtbis(FnHoursBC,lhmin,lhmax,1.0e-8_8,1.0e-10_8,lh) 
		END IF
		IF(ScaleDisutilityIdio==0) llabdisutil = lh**(1.0+1.0/frisch)/(1.0+1.0/frisch)
		IF(ScaleDisutilityIdio==1) llabdisutil = gidioprod * lh**(1.0+1.0/frisch)/(1.0+1.0/frisch)
		
		lc = gbdrift + lh*gnetwage
		IF(lc>0.0) THEN
			lHc = utilfn(lc) - chi*llabdisutil
		ELSE
			lHc = -1.0e12
		END IF
		
		
	ELSE IF	(LaborSupplyGHH==1) THEN
		
		ls = 0.0
		IF(ScaleDisutilityIdio==0) lh = (gnetwage/chi)**(frisch)
		IF(ScaleDisutilityIdio==1) lh = ((gnetwage/gidioprod)/chi)**(frisch)
		
		IF(ImposeMaxHours==1) lh = min(lh, 1.0_8)
		IF(ScaleDisutilityIdio==0) llabdisutil = lh**(1.0+1.0/frisch)/(1.0+1.0/frisch)
		IF(ScaleDisutilityIdio==1) llabdisutil = gidioprod * lh**(1.0+1.0/frisch)/(1.0+1.0/frisch)
		
		lc = gbdrift + lh*gnetwage
		
		IF(lc-chi*llabdisutil>0.0) THEN
			lHc = utilfn(lc-chi*llabdisutil)
		ELSE
			lHc = -1.0e12
		END IF
		
	END IF	
	
	
	
END IF		



END SUBROUTINE OptimalConsumption

File: mycode/OptimalConsumption.m
function [lc,lh,ls,lHc] = OptimalConsumption(lVb,gbdrift,gnetwage,gidioprod)
% set lVb = -999.9 to use budget constraint rather than FOC for consumption
global LaborSupply chi frisch ScaleDisutilityIdio ImposeMaxHours ngpb ngpa utilfn utilfn1 utilfn1inv prefshock

if LaborSupply==0; lh = 1; else; lh = (gnetwage/chi).^frisch; end
if ScaleDisutilityIdio; lh = lh./(gidioprod.^frisch); end
if LaborSupply==1; lh = lh.*lVb.^frisch; end
if ImposeMaxHours; lh = min(lh, 1); end
lc = utilfn1inv(lVb);
ls = gbdrift + lh.*gnetwage - lc;
i = any(lVb <= -999); % stationary point or limit
if any(i(:))
    gbdrift = gbdrift(i);
    gnetwage = repmat(gnetwage,1,ngpb,1);
    gnetwage = gnetwage(i);
    if ScaleDisutilityIdio; gidioprod = squeeze(gidioprod(any(i))); end
    switch LaborSupply
        case 0
            lH = 1;
        case 1
            if ImposeMaxHours; lhmax = 1; else; lhmax = 100; end
            if frisch==1 && prefshock==1
                lhmin = gbdrift./gnetwage/2;
                if ScaleDisutilityIdio
                    lH = sqrt(lhmin.^2+1/chi./gidioprod)-lhmin;
                else
                    lH = sqrt(lhmin.^2+1/chi)-lhmin;
                end
            else
                lhmin = max(0,-gbdrift./gnetwage+1e-5);
                if ScaleDisutilityIdio
                    FnHoursBC = @(lhmin,gbdrift,gnetwage,gidioprod) fzero(@(lh) chi*(lh.^(1/frisch)) - utilfn1(gbdrift + lh.*gnetwage) .* (gnetwage./gidioprod),[lhmin lhmax]);
                    [lH,~,iflag] = arrayfun(FnHoursBC,lhmin,gbdrift,gnetwage,gidioprod);
                else
                    FnHoursBC = @(lhmin,gbdrift,gnetwage) fzero(@(lh) chi*(lh.^(1/frisch)) - utilfn1(gbdrift + lh.*gnetwage) .* gnetwage,[lhmin lhmax]);
                    [lH,~,iflag] = arrayfun(FnHoursBC,lhmin,gbdrift,gnetwage);
                end
                if any(iflag~=1); error('fzero'); end
            end
        case 2        
            lH = (gnetwage/chi).^frisch;
            if ScaleDisutilityIdio; lH = lH./gidioprod.^frisch; end
            if ImposeMaxHours; lH = min(lH, 1); end
    end
    lc(:,i) = repmat(gbdrift + lH.*gnetwage,1,ngpa)';
    lh(:,i) = repmat(lH,1,ngpa)';
    ls(:,i) = 0;
end
llabdisutil = chi*lh.^(1+1/frisch)/(1+1/frisch);
if ScaleDisutilityIdio; llabdisutil = gidioprod.*llabdisutil; end
if LaborSupply==2; ls = ls-llabdisutil; end
lHc = utilfn(lc) + lVb.*ls;
if LaborSupply==1; lHc = lHc-llabdisutil; end
lHc(lc<=0) = -1e12;
if LaborSupply==2; lc = lc+llabdisutil; end